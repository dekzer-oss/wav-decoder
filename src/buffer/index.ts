export * from './AudioBufferPool';
export * from './StreamBuffer';
