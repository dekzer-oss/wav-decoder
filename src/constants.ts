export const INV_128 = 1 / 128;

export const INV_32768 = 1 / 32768;
export const INV_8388608 = 1 / 8388608;
export const INV_2147483648 = 1 / 2147483648;

export const WAVE_FORMAT_EXTENSIBLE = 0xfffe;
export const WAVE_FORMAT_IEEE_FLOAT = 0x0003;
export const WAVE_FORMAT_MULAW = 0x0007;
export const WAVE_FORMAT_PCM = 0x0001;
export const WAVE_FORMAT_IMA_ADPCM = 0x0011;
export const WAVE_FORMAT_ALAW = 0x0006;

export const RIFF_SIGNATURE = 0x46464952 as const;
export const RIFX_SIGNATURE = 0x52494658 as const;
export const WAVE_SIGNATURE = 'WAVE' as const;
export const FMT_CHUNK = 'fmt ' as const;
export const DATA_CHUNK = 'data' as const;

export const IMA_STEP_TABLE = new Int32Array([
  7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 21, 23, 25, 28, 31, 34, 37, 41, 45, 50, 55, 60, 66, 73, 80, 88, 97, 107, 118,
  130, 143, 157, 173, 190, 209, 230, 253, 279, 307, 337, 371, 408, 449, 494, 544, 598, 658, 724, 796, 876, 963, 1060,
  1166, 1282, 1411, 1552, 1707, 1878, 2066, 2272, 2499, 2749, 3024, 3327, 3660, 4026, 4428, 4871, 5358, 5894, 6484,
  7132, 7845, 8630, 9493, 10442, 11487, 12635, 13899, 15289, 16818, 18500, 20350, 22385, 24623, 27086, 29794, 32767,
]);

export const IMA_INDEX_ADJUST_TABLE = new Int8Array([-1, -1, -1, -1, 2, 4, 6, 8, -1, -1, -1, -1, 2, 4, 6, 8]);

export const ALAW_TABLE = (() => {
  const table = new Float32Array(256);
  for (let i = 0; i < 256; i++) {
    let aVal = i ^ 0x55;
    let sign = aVal & 0x80 ? -1 : 1;
    let exponent = (aVal & 0x70) >> 4;
    let mantissa = aVal & 0x0f;
    let sample: number;
    if (exponent === 0) {
      sample = (mantissa << 4) + 8;
    } else {
      sample = ((mantissa + 16) << (exponent + 3)) - 2048;
    }
    table[i] = sign * sample * INV_32768;
  }
  return table;
})();

export const MULAW_TABLE = (() => {
  const MULAW_BIAS = 0x84;
  const table = new Float32Array(256);
  for (let i = 0; i < 256; i++) {
    let muVal = ~i & 0xff;
    let sign = muVal & 0x80 ? -1 : 1;
    let exponent = (muVal & 0x70) >> 4;
    let mantissa = muVal & 0x0f;
    let sample = ((mantissa << 3) + MULAW_BIAS) << exponent;
    sample -= MULAW_BIAS;
    table[i] = sign * sample * INV_32768;
  }
  return table;
})();

export const KSDATAFORMAT_SUBTYPE_IEEE_FLOAT = new Uint8Array([
  0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71,
]);

export const KSDATAFORMAT_SUBTYPE_PCM = new Uint8Array([
  0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71,
]);
