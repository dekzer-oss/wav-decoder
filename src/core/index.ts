export * from './RingBuffer';
export * from './WavDecoder';
export * from './StateManager';
export * from './ErrorFactory';
