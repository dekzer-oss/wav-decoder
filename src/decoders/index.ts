export * as pcm from './pcm';
export * as float from './float';
export * as imaadpcm from './ima-adpcm';
export * as alaw from './a-law';
export * as mulaw from './mu-law';
