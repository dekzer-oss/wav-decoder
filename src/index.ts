export { WavDecoder } from './WavDecoder';
export { WasmDecoder } from './WasmDecoder';
export * from './types';
