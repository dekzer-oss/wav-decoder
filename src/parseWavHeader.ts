import type { DataChunk, WavFormat, WavHeaderParserResult } from './types';
import {
  DATA_CHUNK,
  FMT_CHUNK,
  RIFF_SIGNATURE,
  RIFX_SIGNATURE,
  WAVE_FORMAT_EXTENSIBLE,
  WAVE_FORMAT_IMA_ADPCM,
  WAVE_SIGNATURE,
} from './constants.ts';

export const EMPTY_WAV_HEADER_RESULT: WavHeaderParserResult = {
  isLittleEndian: true,
  format: null,
  isExtensible: false,
  dataBytes: 0,
  dataOffset: 0,
  parsedChunks: [],
  unhandledChunks: [],
  totalSamples: 0,
  totalFrames: 0,
  duration: 0,
  dataChunks: [],
  warnings: [],
  errors: [],
} as const;

export function parseWavHeader(buffer: Uint8Array): WavHeaderParserResult {
  const errors: string[] = [];
  const warnings: string[] = [];
  const len = buffer.length;

  if (len < 12) {
    errors.push('File is too small to be a valid WAV (expected at least 12 bytes)');
    return { ...EMPTY_WAV_HEADER_RESULT, errors, warnings };
  }

  const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);

  let isLittleEndian = false;
  const riffSig = view.getUint32(0, true);
  if (riffSig === RIFF_SIGNATURE) {
    isLittleEndian = true;
  } else {
    const rifxSig = view.getUint32(0, false);
    if (rifxSig === RIFX_SIGNATURE) {
      isLittleEndian = false;
    } else {
      errors.push('Missing RIFF or RIFX signature at byte 0');
      return { ...EMPTY_WAV_HEADER_RESULT, errors, warnings };
    }
  }

  // const fileSize = view.getUint32(4, isLittleEndian);
  // if (fileSize + 8 !== len) {
  //   warnings.push(`Declared file size (${fileSize + 8} bytes) does not match actual buffer size (${len} bytes)`);
  // }

  const waveStr = String.fromCharCode(...buffer.subarray(8, 12));
  if (waveStr !== WAVE_SIGNATURE) {
    errors.push('Missing "WAVE" signature at byte 8');
    return { ...EMPTY_WAV_HEADER_RESULT, errors, warnings };
  }

  let offset = 12;
  let format: WavFormat | null = null;
  const dataChunks: DataChunk[] = [];
  const parsedChunks: DataChunk[] = [];
  const unhandledChunks: DataChunk[] = [];
  let isExtensible = false;

  while (offset + 8 <= len) {
    const chunkStr = String.fromCharCode(...buffer.subarray(offset, offset + 4));
    const chunkSize = view.getUint32(offset + 4, isLittleEndian);
    const paddedSize = chunkSize + (chunkSize & 1);
    const available = len - (offset + 8);
    const actualChunkSize = available < chunkSize ? available : chunkSize;

    parsedChunks.push({ id: chunkStr, offset, size: actualChunkSize });

    if (actualChunkSize < chunkSize) {
      warnings.push(
        `Unexpected end of file in "${chunkStr}" chunk at byte ${offset}. ` +
          `Expected ${chunkSize} bytes, but only ${actualChunkSize} byte${actualChunkSize !== 1 ? 's' : ''} available.`
      );
    }

    if (chunkStr === FMT_CHUNK) {
      if (actualChunkSize < 16) {
        errors.push('"fmt " chunk is too small (expected at least 16 bytes)');
        break;
      }

      const off = offset + 8;
      const formatTag = view.getUint16(off, isLittleEndian);
      const channels = view.getUint16(off + 2, isLittleEndian);
      const sampleRate = view.getUint32(off + 4, isLittleEndian);
      const bytesPerSecond = view.getUint32(off + 8, isLittleEndian);
      const blockAlign = view.getUint16(off + 12, isLittleEndian);
      const bitsPerSample = view.getUint16(off + 14, isLittleEndian);

      let samplesPerBlock: number | undefined;
      let channelMask: number | undefined;
      let validBitsPerSample: number | undefined;
      let subFormat: Uint8Array | undefined;
      let extSize: number | undefined;
      let extraFields: Uint8Array | undefined;

      if (actualChunkSize >= 18) {
        extSize = view.getUint16(off + 16, isLittleEndian);
        const extStart = off + 18;
        const maxEnd = Math.min(off + actualChunkSize, buffer.length);

        if (formatTag === WAVE_FORMAT_EXTENSIBLE && extSize >= 22 && actualChunkSize >= 40) {
          if (extStart + 22 <= maxEnd) {
            isExtensible = true;
            validBitsPerSample = view.getUint16(extStart, isLittleEndian);
            channelMask = view.getUint32(extStart + 2, isLittleEndian);
            subFormat = buffer.slice(extStart + 6, extStart + 22);
          }
        } else if (formatTag === WAVE_FORMAT_IMA_ADPCM && extSize >= 2 && actualChunkSize >= 20) {
          if (extStart + 2 <= maxEnd) {
            samplesPerBlock = view.getUint16(extStart, isLittleEndian);
          }
        }

        const extEnd = extStart + extSize;
        const safeExtEnd = Math.min(extEnd, maxEnd);
        if (safeExtEnd > extStart) {
          extraFields = buffer.slice(extStart, safeExtEnd);
        }
      }

      format = {
        bitsPerSample,
        blockAlign,
        bytesPerSecond,
        channels,
        formatTag,
        sampleRate,
        samplesPerBlock,
        channelMask,
        validBitsPerSample,
        subFormat,
        extSize,
        extraFields,
      };
    } else if (chunkStr === DATA_CHUNK) {
      dataChunks.push({ id: DATA_CHUNK, offset: offset + 8, size: actualChunkSize });
    } else {
      unhandledChunks.push({ id: chunkStr, offset, size: actualChunkSize });
    }

    if (actualChunkSize < chunkSize) break;
    offset += 8 + paddedSize;
  }

  if (!format) {
    errors.push('Missing required "fmt " chunk');
  }

  if (dataChunks.length === 0) {
    warnings.push('Missing "data" chunk â€” no audio payload found');
  }

  const dataBytes = dataChunks.reduce((n, c) => n + c.size, 0);
  const dataOffset = dataChunks[0]?.offset || 0;

  let totalFrames = 0;
  if (format && format.blockAlign > 0) {
    totalFrames = Math.floor(dataBytes / format.blockAlign);
  }
  if (format && format.formatTag === WAVE_FORMAT_IMA_ADPCM && format.samplesPerBlock && format.blockAlign > 0) {
    const numBlocks = Math.floor(dataBytes / format.blockAlign);
    totalFrames = numBlocks * format.samplesPerBlock;
  }

  const totalSamples = format ? totalFrames * format.channels : 0;
  const duration = format && format.sampleRate > 0 ? totalFrames / format.sampleRate : 0;

  return {
    isLittleEndian,
    format,
    isExtensible,
    dataBytes,
    dataOffset,
    parsedChunks,
    unhandledChunks,
    totalSamples,
    totalFrames,
    duration,
    dataChunks,
    warnings,
    errors,
  };
}
