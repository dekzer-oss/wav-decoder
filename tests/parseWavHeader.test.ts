import { parseWavHeader } from '../src/parseWavHeader';
import { describe, expect, it } from 'vitest';
import { WaveFile } from 'wavefile';
import { WAVE_FORMAT_EXTENSIBLE, WAVE_FORMAT_IMA_ADPCM, WAVE_FORMAT_PCM } from '../src/constants';

function createTestBuffer(opts: {
  riffSize?: number;
  chunks: { id: string; size: number; data?: Uint8Array }[];
  endianness?: 'LE' | 'BE';
  corrupt?: { at: number; val: number }[];
}): Uint8Array {
  let finalBufferSize = 12;
  for (const chunk of opts.chunks) {
    finalBufferSize += 8 + (chunk.data ? chunk.data.length : 0);
  }
  const buffer = new Uint8Array(finalBufferSize);
  const view = new DataView(buffer.buffer);
  const isLE = opts.endianness !== 'BE';

  view.setUint32(0, isLE ? 0x46464952 : 0x58464952, true);
  view.setUint32(4, opts.riffSize ?? finalBufferSize - 8, isLE);
  view.setUint32(8, 0x45564157, true);

  let currentOffset = 12;
  for (const chunk of opts.chunks) {
    for (let i = 0; i < 4; i++) buffer[currentOffset + i] = chunk.id.charCodeAt(i);
    view.setUint32(currentOffset + 4, chunk.size, isLE);
    currentOffset += 8;
    if (chunk.data) {
      buffer.set(chunk.data, currentOffset);
      currentOffset += chunk.data.length;
    }
  }
  if (opts.corrupt) for (const { at, val } of opts.corrupt) if (at < buffer.length) buffer[at] = val;
  return buffer;
}

function assertHeader(
  result: ReturnType<typeof parseWavHeader>,
  opts: {
    channels?: number;
    sampleRate?: number;
    bitsPerSample?: number;
    formatTag?: number;
    allowWarnings?: boolean;
    warnings?: Array<RegExp | string>;
    errors?: Array<RegExp | string>;
    allowErrors?: boolean;
  } = {}
) {
  const { allowWarnings = true, allowErrors = false } = opts;

  if (opts.formatTag !== undefined) {
    expect(result.format?.formatTag).toBe(opts.formatTag);
  }
  if (opts.errors) {
    for (const err of opts.errors) {
      if (typeof err === 'string') {
        expect(result.errors).toContain(err);
      } else {
        expect(result.errors.some((e) => err.test(e))).toBe(true);
      }
    }
  } else if (!allowErrors) {
    expect(result.errors).toEqual([]);
  }
}

function snapshotHeader(result: ReturnType<typeof parseWavHeader>) {
  const snap = {
    isLittleEndian: result.isLittleEndian,
    format: result.format
      ? {
          ...result.format,
          extraFields: undefined,
          subFormat: result.format.subFormat ? '[GUID]' : undefined,
        }
      : undefined,
    isExtensible: result.isExtensible,
    dataBytes: result.dataBytes,
    dataOffset: result.dataOffset,
    totalSamples: result.totalSamples,
    totalFrames: result.totalFrames,
    duration: result.duration,
    dataChunks: result.dataChunks.map((c) => ({ offset: c.offset, size: c.size })),
    parsedChunks: result.parsedChunks.map((c) => ({ id: c.id, offset: c.offset, size: c.size })),
    unhandledChunks: result.unhandledChunks.map((c) => ({ id: c.id })),
    errors: result.errors,
  };
  expect(snap).toMatchSnapshot();
}

// For extensible format
const PCM_GUID = new Uint8Array([
  0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71,
]);

describe('parseWavHeader', () => {
  // Table-driven PCM format checks (should always pass)
  it.each([
    { channels: 1, sampleRate: 44100, bits: '16', desc: 'mono 16-bit' },
    { channels: 2, sampleRate: 48000, bits: '24', desc: 'stereo 24-bit' },
    { channels: 1, sampleRate: 44100, bits: '8', desc: 'mono 8-bit' },
    { channels: 2, sampleRate: 96000, bits: '32', desc: 'stereo 32-bit' },
  ])('parses $desc', (fmt) => {
    const wav = new WaveFile();
    wav.fromScratch(fmt.channels, fmt.sampleRate, fmt.bits, [new Int16Array([0, 1, -1, 2, -2, 3, -3])]);
    const result = parseWavHeader(wav.toBuffer());
    assertHeader(result, {
      formatTag: WAVE_FORMAT_PCM,
      channels: fmt.channels,
      sampleRate: fmt.sampleRate,
      bitsPerSample: Number(fmt.bits),
      warnings: [],
      errors: [],
    });
  });

  // Edge cases: Header errors
  it.each([
    ['missing RIFF signature', { corrupt: [{ at: 0, val: 0x00 }], chunks: [] }, { errors: [/RIFF|RIFX/i] }],
    ['missing WAVE signature', { corrupt: [{ at: 8, val: 0x00 }], chunks: [] }, { errors: [/WAVE/i] }],
    [
      'missing fmt chunk',
      { chunks: [{ id: 'data', size: 4, data: new Uint8Array([1, 2, 3, 4]) }] },
      { errors: [/fmt/i] },
    ],
    [
      'truncated fmt chunk',
      {
        chunks: [{ id: 'fmt ', size: 16, data: new Uint8Array(10) }],
      },
      {
        errors: [/truncated in stream/i, /Missing required "fmt "/i],
      },
    ],
  ])('Edge: %s', (_, opts, expectProps) => {
    const buffer = createTestBuffer(opts as any);
    const result = parseWavHeader(buffer);
    assertHeader(result, expectProps as any);
  });

  // Edge cases: Chunks and layout
  it('handles multiple data chunks and unhandled chunks', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 4, data: new Uint8Array([1, 2, 3, 4]) },
        { id: 'junk', size: 8, data: new Uint8Array(8) },
        { id: 'data', size: 2, data: new Uint8Array([5, 6]) },
      ],
    });
    const result = parseWavHeader(buffer);
    // This triggers 0 channels/sample rate/bits per sample errors due to zero-initialized fmt data
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.dataChunks.length).toBe(2);
    expect(result.dataBytes).toBe(6);
    expect(result.unhandledChunks.map((c) => c.id)).toContain('junk');
  });

  it('handles chunk with odd size (padding)', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 5, data: new Uint8Array([1, 2, 3, 4, 5]) },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.dataChunks[0]!.size).toBe(5);
    expect(result.dataBytes).toBe(5);
  });

  it('handles junk chunk before fmt', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'junk', size: 8, data: new Uint8Array(8) },
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 4 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.unhandledChunks.map((c) => c.id)).toContain('junk');
    expect(result.format).toBeDefined();
  });

  // Edge: file size warnings
  it('warns when RIFF size < actual size', () => {
    const buffer = createTestBuffer({
      riffSize: 20,
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 8 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
  });

  it('warns when RIFF size > actual size', () => {
    const buffer = createTestBuffer({
      riffSize: 100,
      chunks: [{ id: 'fmt ', size: 16, data: new Uint8Array(16) }],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
  });

  it('handles truncated buffer (partial last chunk)', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 100, data: new Uint8Array(50) },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.dataBytes).toBe(0);
  });

  // Edge: Non-PCM/extensible/ADPCM
  it('parses extensible format with channel mask', () => {
    const fmtData = new Uint8Array(40);
    const view = new DataView(fmtData.buffer);
    view.setUint16(0, WAVE_FORMAT_EXTENSIBLE, true);
    view.setUint16(2, 2, true);
    view.setUint32(4, 48000, true);
    view.setUint32(8, 192000, true);
    view.setUint16(12, 8, true);
    view.setUint16(14, 32, true);
    view.setUint16(16, 22, true);
    view.setUint16(18, 24, true); // valid bits
    view.setUint32(20, 0x3, true); // channel mask
    fmtData.set(PCM_GUID, 24);
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 40, data: fmtData },
        { id: 'data', size: 100 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      formatTag: WAVE_FORMAT_EXTENSIBLE,
      errors: [/truncated in stream/i, /bytes per second mismatch/i],
    });
    expect(result.isExtensible).toBe(true);
    expect(result.format?.validBitsPerSample).toBe(24);
    expect(result.format?.channelMask).toBe(0x3);
    expect(result.format?.subFormat).toEqual(PCM_GUID);
  });

  it('parses IMA ADPCM format', () => {
    const fmtData = new Uint8Array(20);
    const view = new DataView(fmtData.buffer);
    view.setUint16(0, WAVE_FORMAT_IMA_ADPCM, true);
    view.setUint16(2, 1, true);
    view.setUint32(4, 44100, true);
    view.setUint32(8, 22050, true);
    view.setUint16(12, 512, true);
    view.setUint16(14, 4, true);
    view.setUint16(16, 2, true);
    view.setUint16(18, 505, true);
    const dataChunkData = new Uint8Array(1024);
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 20, data: fmtData },
        { id: 'data', size: 1024, data: dataChunkData },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      formatTag: WAVE_FORMAT_IMA_ADPCM,
      errors: [/non-byte-aligned bits per sample/i, /bytes per second mismatch/i],
    });
    // Spec math
    const numBlocks = Math.floor(1024 / 512);
    const totalFrames = numBlocks * 505;
    const totalSamples = totalFrames * 1;
    expect(result.totalSamples).toBe(totalSamples);
  });

  // Zero/unusual
  it('handles zero sample rate', () => {
    const fmtData = new Uint8Array(16);
    const view = new DataView(fmtData.buffer);
    view.setUint16(0, WAVE_FORMAT_PCM, true);
    view.setUint16(2, 2, true);
    view.setUint32(4, 0, true);
    view.setUint16(14, 16, true);
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: fmtData },
        { id: 'data', size: 100 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 Hz sample rate/],
    });
    expect(result.duration).toBe(0);
  });

  it('handles empty data chunk', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 0 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.dataBytes).toBe(0);
    expect(result.totalSamples).toBe(0);
  });

  it('handles minimal valid WAV', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 0 },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
  });

  // RIFX / big-endian
  it('parses big-endian header', () => {
    const buffer = createTestBuffer({
      endianness: 'BE',
      chunks: [
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
        { id: 'data', size: 4, data: new Uint8Array([1, 2, 3, 4]) },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.format).toBeDefined();
    expect(result.dataBytes).toBe(4);
  });

  it('allows data chunk before fmt ', () => {
    const buffer = createTestBuffer({
      chunks: [
        { id: 'data', size: 4, data: new Uint8Array([1, 2, 3, 4]) },
        { id: 'fmt ', size: 16, data: new Uint8Array(16) },
      ],
    });
    const result = parseWavHeader(buffer);
    assertHeader(result, {
      errors: [/Invalid format: 0 channels/, /Invalid format: 0 Hz sample rate/, /Invalid format: 0 bits per sample/],
    });
    expect(result.format).toBeDefined();
    expect(result.dataChunks.length).toBe(1);
  });

  it('structure sanity snapshot (basic PCM)', () => {
    const wav = new WaveFile();
    wav.fromScratch(2, 44100, '16', [new Int16Array([0, 1, -1, 2, -2, 3, -3])]);
    const result = parseWavHeader(wav.toBuffer());
    snapshotHeader(result);
  });
});
