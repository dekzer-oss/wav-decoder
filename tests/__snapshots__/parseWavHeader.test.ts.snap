// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseWavHeader > structure sanity snapshot (basic PCM) 1`] = `
{
  "dataBytes": 14,
  "dataChunks": [
    {
      "offset": 44,
      "size": 14,
    },
  ],
  "dataOffset": 44,
  "duration": 0.00006802721088435374,
  "errors": [],
  "format": {
    "bitsPerSample": 16,
    "blockAlign": 4,
    "bytesPerSecond": 176400,
    "channelMask": undefined,
    "channels": 2,
    "extSize": undefined,
    "extraFields": undefined,
    "formatTag": 1,
    "sampleRate": 44100,
    "samplesPerBlock": undefined,
    "subFormat": undefined,
    "validBitsPerSample": undefined,
  },
  "isExtensible": false,
  "isLittleEndian": true,
  "parsedChunks": [
    {
      "id": "fmt ",
      "offset": 12,
      "size": 16,
    },
    {
      "id": "data",
      "offset": 36,
      "size": 14,
    },
  ],
  "totalFrames": 3,
  "totalSamples": 6,
  "unhandledChunks": [],
}
`;
